src/ldeboot.c

```diff
@@ -200,12 +200,19 @@ int main(int argc, char *argv[]) {
     argv[0] = filetorun;
   } else {
     /* copy up to and including the final "/" in the path */
-#ifndef MAIKO_OS_HAIKU 
-    dirsepp = stpncpy(filetorunpath, argv[0], dirsepp + 1 - argv[0]);
-#else
-    dirsepp = strncpy(filetorunpath, argv[0], dirsepp + 1 - argv[0]);
+
+#ifdef MAIKO_OS_HAIKU
+    /* Copied from libiberty stpncpy.c */
+    char * stpncpy (char *dst, const char *src, size_t len) {
```

Comment:
Umm...  function definition is not allowed here.  What C compiler are you using??

Reply:
Am i correct in assuming that the original problem in this area was that Haiku doesn't provide a definition for `stpncpy()` in libroot?  If it's present, but in another library, link against that.  If it isn't present at all, also file a bug report against either Haiku itself (to get stpncpy implemented), or the libroot documentation that suggests it implements POSIX requirements but doesn't, and doesn't provide a list of unimplemented functions (stpncpy is a POSIX function).

Reply:
Indeed it appears Haiku is missing `stpncpy`. However, it's an "Issue 7" (2018) addition, and Haiku's `_POSIX_VERSION` is `200809L`, meaning we don't make any guarantees about compliance past then. (We have implemented a lot of POSIX-2018 but not all of it yet.)

I will work on getting `stpncpy` added to Haiku's `libroot`.

Reply:
>  What C compiler are you using??

Inline function definition like this is a GCC extension, IIRC.

Reply:
https://gcc.gnu.org/onlinedocs/gcc/Nested-Functions.html

Reply:
Yeah, ok.  We try to stick to ISO/ANSI C.  I ported the code forward from K&R C to C89, and then we've started to use C99 in some cases, but ISO C forbids nested functions.  Some GCC extensions have crept in (e.g., computed goto in the `-DOPDISP` case) but we try to keep them at an absolute minimum to preserve portability/compatibility with older compilers and systems.

Reply:
@waddlesplash `stpncpy()` is present in The Open Group Base Specifications Issue 7, IEEE Std 1003.1-2008, which says _POSIX_C_SOURCE has value 200809L, so I'm surprised it's not there.

Reply:
@nbriggs I can't find it in the 2008 edition. Do you have a link? I can only find it in the 2018.

Reply:
@waddlesplash - https://pubs.opengroup.org/onlinepubs/9699919799.2008edition/ - but search seems to shift to a more recent edition, so here's the direct link - https://pubs.opengroup.org/onlinepubs/9699919799.2008edition/functions/stpncpy.html

But thanks for pressing on with adding it to the Haiku library.